packages/artifact/src/internal-download-http-client.ts

```diff
@@ -12,122 +12,174 @@ import {
   QueryArtifactResponse
 } from './internal-contracts'
 import {IHttpClientResponse} from '@actions/http-client/interfaces'
-import {HttpClient} from '@actions/http-client'
+import {HttpManager} from './internal-http-manager'
 import {DownloadItem} from './internal-download-specification'
-import {getDownloadFileConcurrency} from './internal-config-variables'
+import {
+  getDownloadFileConcurrency,
+  getRetryWaitTime
+} from './internal-config-variables'
 import {warning} from '@actions/core'
 
-/**
- * Gets a list of all artifacts that are in a specific container
- */
-export async function listArtifacts(): Promise<ListArtifactsResponse> {
-  const artifactUrl = getArtifactUrl()
-  const client = createHttpClient()
-  const requestOptions = getRequestOptions('application/json')
+export class DownloadHttpClient {
+  private downloadHttpManager: HttpManager
 
-  const rawResponse = await client.get(artifactUrl, requestOptions)
-  const body: string = await rawResponse.readBody()
-  if (isSuccessStatusCode(rawResponse.message.statusCode) && body) {
-    return JSON.parse(body)
+  constructor() {
+    this.downloadHttpManager = new HttpManager()
   }
-  // eslint-disable-next-line no-console
-  console.log(rawResponse)
-  throw new Error(`Unable to list artifacts for the run`)
-}
 
-/**
- * Fetches a set of container items that describe the contents of an artifact
- * @param artifactName the name of the artifact
- * @param containerUrl the artifact container URL for the run
- */
-export async function getContainerItems(
-  artifactName: string,
-  containerUrl: string
-): Promise<QueryArtifactResponse> {
-  // The itemPath search parameter controls which containers will be returned
-  const resourceUrl = new URL(containerUrl)
-  resourceUrl.searchParams.append('itemPath', artifactName)
+  /**
+   * Gets a list of all artifacts that are in a specific container
+   */
+  async listArtifacts(): Promise<ListArtifactsResponse> {
+    const artifactUrl = getArtifactUrl()
+    this.downloadHttpManager.createClients(1)
```

Comment:
This is unnecessary - what you are doing here is just create a single http client.  Involving the `downloadHttpManager` here complicates the logic and hide your intention.  Also this `downloadHttpManager` is a singleton in this class, so if the consumer calls `listArtifacts()` twice on async threads, your http clients may be corrupted - thread b is creating the http client, and thread a is getting the client in line 36.

Reply:
The user won't be able to call `listArtifacts` twice because it's an internal method. What we publicly expose is `uploadArtifact()`, `downloadAritfact()` and `downloadAllArtifacts()`. These are the methods that are exposed as part of the public interface for an artifactClient after `create()` gets called. They could poke around and try to call these methods themselves, but that's a large reason as to why `internal` is used in the naming everywhere so it's obvious what is public and what isn't.

Whenever `downloadArtifact` is called, a new downloadhttpClient class is created, right [here](https://github.com/actions/toolkit/blob/640fc522a7d780531cd610c1d473829a4d237e8a/packages/artifact/src/internal-artifact-client.ts#L140) so there should not be any problems with corruption. When a new class gets created, a new manager gets created as part of the constructor [here](https://github.com/actions/toolkit/blob/640fc522a7d780531cd610c1d473829a4d237e8a/packages/artifact/src/internal-download-http-client.ts#L27).

I did some refactoring to make this a class so it's easier to maintain state and not run into any problems if `downloadArtifact()` is called multiple times in quick succession. Each one of these classes has their own `downloadHttpManager` (or `uploadHttpManager` if uploading). I experimented with a singleton earlier on but decided against it precisely because there could be problems if `uploadArtifact` or `download` artifact get called multiple times in quick succession

Reply:
Okay, as long as we don't call dispose on those clients then it's good.